### ISSUE TITLE
Support ASP.NET Core

### ISSUE THREAD

**Issue by klings** [NWebsec/NWebsec#59]
Rumor has it that something's cooking in Redmond. NWebsec should be updated to run on the new frameworks.

**Comment by vossad01** [comment]
`NWebsec.Owin` middleware can be used in ASP.NET 5 (I am using it)

[A simple adapter method is needed](http://stackoverflow.com/a/30742029/1072626) and then you can register the components in ASP.NET 5.  Packaging up some extension methods based on that would be a fairly simple way to make it *look* native to ASP.NET 5 users.

There is separate, but similar issue of supporting .the 'NET Core' platform which will be required for anyone wanting to develop new ASP.NET 5 application for that runtime.

**Comment by klings** [comment]
Thanks for leaving a note. Yes, I expected that it should be fairly straightforward to get the OWIN middleware up and running on the new version. I took care to build the middleware according to the spec, without any dependencies on the MS stuff introduced in Katana (contexts etc).

But AFAIK, I'll have to build new MVC attributes as those are dependent on the HttpContext in the "old" version so that's where I'll have to put in a few hours of work.

Great to hear that NWebsec is already partially up and running on vnext!

**Comment by RehanSaeed** [comment]
Yes, they've totally changed the filter attributes. I've been looking into this too. I'm amazed at how much is missing in MVC 6 so far.

**Comment by klings** [comment]
Well, it seems they're hard at work to fill in all the missing pieces. They're landing a lot of big decisions these days, with the rename to dnx and now with the refactoring of the (MVC) packages. Many important base packages are also surfacing now for the core framework, e.g. System.Security...X590Certificates which will be needed for the HPKP functionality.

I've decided to keep an eye on what they're doing, but hang back for a while before I start coding. When things stabilize, I'll get to work.

**Comment by singlewind** [comment]
beta7 will be a good start version. Because most api will be done consolidated. beta8 will be the last feature fill version. They have released the roadmap for it. At the moment, the stable version is beta6 and the dev version is beta8. beta7 will be released end of August. It is just waiting other team to catch up. It is amazing all the teams work together to re-implement the web part of .NET

**Comment by millusdk** [comment]
Any updates on this? :)

**Comment by RehanSaeed** [comment]
Beta 8 is out next week I think. I'd wait until that happens before starting any upgrade. Beta 8 will be feature complete, only bugs left to fix. A good time to start a port.

**Comment by klings** [comment]
I haven't done any concrete changes yet, support for ASP.NET is "under planning". There's a couple of new features I'd like to get out the door first. Also, with beta 8 so close I figured I'd wait until we had the feature complete bits available.

**Comment by klings** [comment]
So, the port to dnx is coming along nicely, targeting dnx451 is the first goal. The NWebsec projects rely on NUnit and Moq, unfortunately they don't (fully) support .NET Core just yet.

nunit/nunit#927 - unit tests can be run from command line
Moq/moq4#168 - still missing .NET Core support

We'll see how it goes, but this might delay .NET Core support in NWebsec for a while. I'll anyhow have to rewrite a few pieces of the code that's not compatible with .NET core, so I'll start there.

**Comment by RehanSaeed** [comment]
I made the switch to xUnit which has better support. NUnit support is on it's way but right now does not work with Visual Studio. Can't help with Moq.

**Comment by oshalygin** [comment]
@RehanSaeed yeah but the more I use xUnit the more I stay glued to NUnit and this repo.  Not a fan of the ideology @ xUnit...

**Comment by RehanSaeed** [comment]
I have code using both at work and need to make a decision to stick with one. xUnit wins by default stop far due to .xproj unit testing support. @oshalyn I'd be interested by what you mean by ideology?

**Comment by klings** [comment]
NUnit support will come along, but right now it's a bit cumbersome to use it. I'm running the tests through console apps for now. Really looking forward to the tooling catching up with the dnx projects.

**Comment by klings** [comment]
So, I've pushed a Gamma release of the "new" NWebsec just now. There's been a restructuring of the packages, there are now three:

* NWebsec.Core
* NWebsec.Middleware
* NWebsec.Mvc

These are all ASP.NET 5 only. I'll continue working towards an RC, but there'll be some minor breaking changes on the way. This is a gamma-release because the code has simply been ported to the new framework, and is safe to use in production.

**Comment by RehanSaeed** [comment]
Whats the best way to go from the code below in Global.asax.cs to ASP.NET 5 since Global.asax.cs in ASP.NET 5 still exists if you really want it but using it is discouraged.

I was thinking of just adding an action to my HomeController but the `CspViolationReport` and `CspReportDetails` classes seem to be missing. Thought I'd ask before I go down a rabbit hole.

```
protected void NWebsecHttpHeaderSecurityModule_CspViolationReported(object sender, CspViolationReportEventArgs e)
{
    CspViolationReport violationReport = e.ViolationReport;
    CspReportDetails reportDetails = violationReport.Details;
}
```

**Comment by RehanSaeed** [comment]
I found another issue. I need to add `data:` as an CspImgSrcAttribute to get BrowserLink working but I get the following exception:

```
System.ArgumentException was unhandled by user code
  HResult=-2147024809
  Message=CspImgSrcAttribute: Invalid source(s): data:
Details: IDN labels must be between 1 and 63 characters long.
Parameter name: unicode
  Source=NWebsec.Mvc
  StackTrace:
       at NWebsec.Mvc.HttpHeaders.Csp.Internals.CspDirectiveAttributeBase.set_CustomSources(String value)
       at WebApplication3.Startup.ConfigureContentSecurityPolicyFilters(IHostingEnvironment environment, ICollection`1 filters) in C:\Test\WebApplication3\WebApplication3\Startup.Filters.cs:line 276
       at WebApplication3.Startup.<>c__DisplayClass10_0.<ConfigureServices>b__1(MvcOptions mvcOptions) in C:\Test\WebApplication3\WebApplication3\Startup.cs:line 101
       at Microsoft.Extensions.OptionsModel.ConfigureOptions`1.Configure(TOptions options)
       at Microsoft.Extensions.OptionsModel.OptionsManager`1.<>c.<get_Value>b__4_0(TOptions options, IConfigureOptions`1 setup)
       at System.Linq.Enumerable.Aggregate[TSource,TAccumulate](IEnumerable`1 source, TAccumulate seed, Func`3 func)
       at Microsoft.Extensions.OptionsModel.OptionsManager`1.get_Value()
       at Microsoft.AspNet.Mvc.ApplicationModels.DefaultApplicationModelProvider..ctor(IOptions`1 mvcOptionsAccessor)
  InnerException: 
```

**Comment by klings** [comment]
Yes, There's currently no CSP report handling in the middleware, this was also lacking in the Katana-middleware. My thoughts are to add a report handling middleware, which also can be used standalone. Also, I need to make it play nicely with the CSP middleware. This is one of the reasons it is a gamma release, and not an rc.

The data: things looks to be a bug, I'll investigate. Thanks for reporting this!

**Comment by klings** [comment]
My bad on the `data:` source. I've pushed new packages that resolve this.

**Comment by RehanSaeed** [comment]
Added NWebSec as an optional feature to ASP.NET MVC Boilerplate. It's turned on by default.

**Comment by PinpointTownes** [comment]
@klings hey. Do you have plans for ASP.NET Core RC2? I'd love to help if needed.

**Comment by klings** [comment]
Yep, I have gotten hold of the new version and will have a look at it this weekend. I assume that it's pretty straight forward to port to the new version. I'd like to take a stab at this myself, to get to know the changes they've done in RC2. Also, there'll be a restructuring of the NWebsec packages.

Thanks for offering your help, I'll let you know if it's needed.

**Comment by klings** [comment]
@PinpointTownes So, I've pushed the first version of the RC 2 packages to NuGet. They're still gamma, I still have some cleanup to do after porting the code from the "old" ASP.NET. Let me know if you run into any problems.

There are brand new .NET Core packages now, named NWebsec.AspNetCore.*. All namespaces are updated accordingly, other than that there are no API changes.

**Comment by PinpointTownes** [comment]
> There are brand new .NET Core packages now, named NWebsec.AspNetCore.*

Core all the things :trollface: 

Quick question: is there any reason to require `.NET Standard 1.5` instead of just `.NET Standard 1.4`?

**Comment by klings** [comment]
I had a look at [Microsoft.AspNetCore.Mvc](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc/1.0.0-rc2-final) and 1.5 seemed to be the baseline. I've been sifting through some of the other MSFT AspNetCore packages on NuGet now, and see that most of them target `.NETStandard 1.3` for some reason.

1.5 seemed like a reasonable target platform, since .NET Core 1.0 maps to it: [Mapping the .NET Platform Standard to platforms](https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-platform-standard.md#mapping-the-net-platform-standard-to-platforms).

I assume you're targeting 1.4 in your project? Any particular reason for that?

**Comment by PinpointTownes** [comment]
> I assume you're targeting 1.4 in your project? Any particular reason for that?

No particular reason, I just try to use the lowest TFM when possible since there's no reason to require a higher framework when it's not necessary (`netstandard1.3` is probably the most common one in ASP.NET Core for this exact reason).

Here, I guess NWebSec Core/Middleware could target `netstandard1.3` as the only dependency they need is `Microsoft.AspNetCore.Http.Abstractions`. Also, referencing `NETStandard.Library` shouldn't be necessary (and probably an overkill, as this metapackage references a ton of other packages you probably don't need).

**Comment by klings** [comment]
Yup, the `netstandard` library was on my list for investigation as that was new for RC2. I've trimmed the dependencies to only those needed now. It seems several other package authors are also reading up on the new stuff and trying to navigate the new landscape. That's the price we pay for not keeping up with the nightly builds. :)

I'll revisit the `netstandard` version, it might make sense to differentiate between the NWebsec libraries. If you have any concrete examples of where this would be beneficial that'd be much appreciated. If not, I lean towards just keeping the libraries in sync.

In any case, this seems to be a [moving target](https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-platform-standard.md#mapping-the-net-platform-standard-to-platforms). They've updated the matrix, .NET Core 1.0 now maps to `netstandard1.6`, so I assume there will be changes yet again when we get to RTM...

**Comment by RehanSaeed** [comment]
Thanks for the heads up about 1.6, I'd missed that. I guess that the missing `SmtpClient` and `SyndicationFeed` API's will end up in 1.6.

**Comment by RehanSaeed** [comment]
The endless upgrades are finally coming to a close with ASP.NET Core 1.0. This upgrade is fairly trivial in my experience and the only files you are likely needing to change are global.json and project.json. If you don't have time to upgrade, I'll submit a PR at some point.

**Comment by klings** [comment]
I'll update very soon, I see that things have stabilized. Yesterday I had to prioritize the GoT season finale. :)

**Comment by RehanSaeed** [comment]
Cool, lol!

**Comment by klings** [comment]
I pushed new packages just now, built against 1.0.

**Comment by RehanSaeed** [comment]
Any plans to update for 1.1?

**Comment by RehanSaeed** [comment]
1.1.1 is out.

**Comment by klings** [comment]
Pushed new packages @RehanSaeed. They target the ASP.NET versions (1.0.x) included with current new projects from the cli or VS 2015/2017, but work with the 1.1 packages. I'll up the dependencies, upgrade the project files and push 1.0 packages when VS 2017 is released in a week or so.

**Comment by klings** [comment]
v 1.0 is out! :)

**Comment by RehanSaeed** [comment]
Great stuff!
